keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,308 @@
+
+### Non-Goals
+- Additional qos design
+- Support other resources qos
+- Consider QOSReserved feature
+
+## Proposal
+Use memory controller of cgroups v2 to support memory qos for guaranteeing pod/container memory requests/limits and node resource.
+
+Currently we only use `memory.limit_in_bytes=sum(pod.spec.containers.resources.limits[memory])` with cgroups v1 and `memory.max=sum(pod.spec.containers.resources.limits[memory])` with cgroups v2 to limit memory usage. `resources.requests[memory]` is not yet used neither by cgroups v1 nor cgroups v2 to protect memory requests. About memory protection, we use `oom_scores` to determine order of killing container process when OOM occurs. Besides, kubelet can only reserve memory from node allocatable at node level, there is no other memory protection for node resource.
+
+So there are missing some memory protection, it may cause:
+- Pod/Container memory requests can't be fully reserved, page cache is at risk of being recycled
+- Pod/Container memory allocation is not well protected, there may occur allocation latency frequently when node memory nearly runs out
+- Memory overcommit of container is not throttled, there may increase risk of node memory pressure
+- Memory resource of node can't be fully retained and protected
+
+Cgroups v2 introduces a better way to protect and guarantee memory quality.
+
+| File | Description |
+| -------- | -------- |
+| memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. |
+| memory.max | memory.max is the memory usage hard limit, acting as the final protection mechanism: If a cgroup's memory usage reaches this limit and can't be reduced, the system OOM killer is invoked on the cgroup. Under certain circumstances, usage may go over the memory.high limit temporarily. When the high limit is used and monitored properly, memory.max serves mainly to provide the final safety net. The default is max. |
+| memory.low | memory.low is the best-effort memory protection, a "soft guarantee" that if the cgroup and all its descendants are below this threshold, the cgroup's memory won't be reclaimed unless memory can’t be reclaimed from any unprotected cgroups. |
+| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup's memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. |
+
+This proposal uses `memory.min` to protect memory requests for pods and containers. Memory limits are throttled by `memory.high` for `Burstable` containers. We also set `memory.min` for qos and node if corresponding flags are satisfied.
```

Comment:
i think qos is less pertinent to this discussion.

we have cases where memory.request is set (or not)
we have cases where memory limit is set (or not)

i think i am biased to saying the following:

if memory.request is set, that maps to `memory.min` and that aligns to the schedulable guarantee.
if memory.request is not set, we do not set `memory.min`

if memory.limit is set, that maps to `memory.max` and that aligns with cgroups v1 memory.limit in bytes
if memory.limit is not set, we could set `memory.max` to match node allocatable.

to leverage `memory.high`, i think i prefer to introduce a throttling factor.

if memory.limit is set, we set `memory.high` to `memory.limit * throttling.factor` and ensure `memory.high` is always greater than `memory.min`.

if memory.limit is not set, we set `memory.high` to `node allocatable memory * throttling factor`

i know @mrunalp and i wanted to meet with red hat kernel team to understand if there was any issues with `memory.min` being equal to `memory.max`, but in the case of qos=Guaranteed, by the heuristic above, we would omit setting `memory.high`.

Reply:
updated.